Move to Gorilla Mux

pauln opened the pull request:
As discussed in #209, moving to Gorilla Mux provides straightforward support for different HTTP verbs on the same path (thereby allowing for RESTful APIs to be generated).

Closes #209.

pauln committed:
Move to Gorilla Mux in order to support mutliple HTTP verbs (i.e. GET and POST) on the same path

Gorilla Mux also handles path params for us, so we no longer need to process them ourselves.

pauln committed:
Add "LouderGet" RPC to _example/echo.proto to demo multiple HTTP verbs on the same path

pauln committed:
Update tests to match what we expect after moving to Gorilla Mux

Removed tests for funcs which have been removed; updated expected output from GenServerDecode

adamryman committed:
Add github.com/gorilla/mux to vendor

adamryman commented:
Keep going!

https://github.com/pauln/truss/pull/1

pauln committed:
Merge pull request #1 from adamryman/gorilla-mux

Add github.com/gorilla/mux to vendor

adamryman commented:
Question: in `$GOPATH/src/github.com/TuneLab/truss` does `make test` work? Not sure about windows.

This runs the integration tests, to make sure http transport works as we expect.

See here: https://github.com/TuneLab/truss/blob/master/DEVELOPING.md

pauln commented:
Unfortunately, `make test` doesn't work on Windows.  I might just switch to my Linux VM for dealing with the integration tests, as I don't really have time to try to port `make test` to Windows at the moment.

pauln committed:
Use upper-case HTTP verbs in generated HTTP client

pauln commented:
The bulk of the failures were due to the generated HTTP requests using lower-case verbs.  Running them through ToUpper gets a lot further through the integration tests, but `TestGetWithPathParamsRequest_IncompletePath` fails as it's no longer relevant (as written).  The test is expecting a JSON error complaining about the number of parts in the path, but the malformed request now simply results in an HTTP 404 (Not Found).

A couple of options immediately spring to mind:

- Remove this test entirely, as it's now effectively testing how Gorilla Mux handles a (specific) bad request
- Update this test to check for a 404 response

If you let me know which approach (whether one of these or something else entirely) you'd prefer, I'm happy to implement it.

adamryman commented:
Unless you want to re-implement it, I say that it is fine to remove it entirely as I believe that gorilla mux would already test this.

This is awesome work, we have been hoping to get around to this. :100: 

I'll do a real review tomorrow morning. Looks great at a glance though. Should definitely be able to get this in before end of day tomorrow.

adamryman commented on gengokit/httptransport/templates/server.go:
```diff
@@ -49,79 +46,6 @@ var ServerDecodeTemplate = `
 // to fail. So don't change them purely because you think they look a little
 // funny.
 
-// PathParamsTemplate is a source code literal of the code for the PathParams
```
It is great to see this code removed, maybe @hasLeland feels sad about it because he wrote it, though he is probably glad that we don't have to maintain it anymore.

hasLeland replied:
More code that we don't have to maintain is always an improvement!

pauln committed:
Remove TestGetWithPathParamsRequest_IncompletePath

This test is now redundant, as it's testing behaviour which should already be tested sufficiently by Gorrila Mux

pauln commented:
> I believe that gorilla mux would already test this

My thoughts exactly - consider it removed!

adamryman reviewed (approved):
I lied, I just reviewed it now.

So simple, such a small actual diff. Mostly removing code. (besides vendoring) :1st_place_medal: 

@pauln you are being talked about in the TuneLab office, we are excited to have a new contributer adding the features they care about.

   LGTM
 \ʕ◔ϖ◔ʔ/

(Besides either removing or rewriting the one failing integration test)

adamryman commented:
THEY PASS. :shipit:
